### PR TITLE
initial support for npe2 widget

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
experimental support for dock widgets from npe2

along with https://github.com/tlambert03/npe2/pull/26

### CODE REVIEW

File: napari/_qt/menus/plugins_menu.py [hunk 1]
```diff
@@ -51,48 +51,71 @@ def _remove_unregistered_widget(self, event):
...
+        try:
+            import npe2
+        except ImportError:
+            _iterable = iter([])
+        else:
+            pm = npe2.PluginManager.instance()
+            wdgs: DefaultDict[str, List[str]] = DefaultDict(list)
+            for wdg_contrib in pm.iter_widgets():
+                wdgs[wdg_contrib.plugin_name].append(wdg_contrib.name)
+            _iterable = (('dock', x) for x in wdgs.items())
```
Comment (tlambert03) [inline napari/_qt/menus/plugins_menu.py:70]: this is the only new section here... the rest is just refactored to reduce code complexity

**Review by sofroniewn — APPROVED** [review]
LGTM!

File: napari/_qt/menus/plugins_menu.py [hunk 2]
```diff
@@ -51,48 +51,71 @@ def _remove_unregistered_widget(self, event):
...
-                    self.addMenu(menu)
-                else:
-                    menu = self
+        from ...plugins import plugin_manager
 
-                for wdg_name in widgets:
-                    key = (plugin_name, wdg_name)
-                    if multiprovider:
-                        action = QAction(wdg_name, parent=self)
+        _iterable: Iterator[Tuple[str, Tuple[str, Sequence[str]]]]
```
Comment (nclack) [inline napari/_qt/menus/plugins_menu.py:60]: Can you add a comment describing a typical iterator element? Maybe just an example.

File: napari/_qt/menus/plugins_menu.py [hunk 3]
```diff
@@ -51,48 +51,71 @@ def _remove_unregistered_widget(self, event):
...
+        ):
+            if call_all or event.value == plugin_name:
+                self._add_plugin_actions(hook_type, plugin_name, widgets)
+
+    def _add_plugin_actions(
+        self, hook_type: str, plugin_name: str, widgets: Sequence[str]
+    ):
+        from ...plugins import menu_item_template
+
+        print(plugin_name, widgets)
```
Comment (nclack) [inline napari/_qt/menus/plugins_menu.py]: leftover print? Maybe make into a `debug` or `trace` message.

**Review by nclack — APPROVED** [review]

File: napari/_qt/menus/plugins_menu.py (hunk 3 above)
Comment (tlambert03) [reply napari/_qt/menus/plugins_menu.py]: thanks, leftover.  prefer to delete (don't love having logger.debug all over from the vestiges of the initial implementation)

File: napari/_qt/menus/plugins_menu.py (hunk 2 above)
Comment (tlambert03) [reply napari/_qt/menus/plugins_menu.py:60]: added

**Pull request merged**